internal/provider/resource_gitlab_runner.go

```diff
@@ -0,0 +1,257 @@
+package provider
+
+import (
+	"context"
+	"fmt"
+	"log"
+	"strconv"
+
+	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/validation"
+	"github.com/xanzy/go-gitlab"
+)
+
+var _ = registerResource("gitlab_runner", func() *schema.Resource {
+	return &schema.Resource{
+		Description: `The ` + "`gitlab_runner`" + ` resource allows registering a runner, either at an instance level
+or at a group level. The runner will be registered at a group level if the token used is from a group, or at an 
+instance level if the token used is for the instance.
+
+**Upstream API**: [GitLab REST API docs](https://docs.gitlab.com/ee/api/runners.html#register-a-new-runner)`,
+
+		// Since this resource updates an in-place resource, the update method is the same as the create method
+		CreateContext: resourceGitLabRunnerCreate,
+		UpdateContext: resourceGitLabRunnerUpdate,
+		ReadContext:   resourceGitLabRunnerRead,
+		DeleteContext: resourceGitLabRunnerDelete,
+		Importer: &schema.ResourceImporter{
+			StateContext: schema.ImportStatePassthroughContext,
+		},
+
+		Schema: map[string]*schema.Schema{
+			"token": {
+				Description: `The registration token used to register the runner`,
+				Type:        schema.TypeString,
+				ForceNew:    true,
+				Required:    true,
+				Sensitive:   true,
+			},
+			// Even though the output variable is just called 'token', we need to differentiate between the regitration
+			// and authentication token in terraform.
+			"authentication_token": {
+				Description: `The authentication token used for building a config.toml file. This value is not present when imported.`,
+				Type:        schema.TypeString,
+				Computed:    true,
+				Sensitive:   true,
+			},
+			"status": {
```

Comment:
NIT: Let's have the computed-only fields at the end of the schema definition for more readability :)